share/man/man4/ace2.4
.\" Copyright (c) 2025 John Baldwin <john@araratriver.co>
.\"
.\" SPDX-License-Identifier: BSD-2-Clause
.\"
.\" This software was developed by SRI International, the University of
.\" Cambridge Computer Laboratory (Department of Computer Science and
.\" Technology), and Capabilities Limited under Defense Advanced Research
.\" Projects Agency (DARPA) Contract No. FA8750-24-C-B047 ("DEC").
.\"
.Dd February 5, 2026
.Dt ACE2 4
.Os
.Sh NAME
.Nm ace2
.Sh DESCRIPTION
This module provides a set of character devices which provide utilities used
in the evaluation of the DARPA CPM program.
.Pp
The
.Nm
module provides the following character devices in
.Pa /dev :
.Bl -tag -width "ace2-ace-capability"
.It Pa ace2-ace-data
Permits reading and writing arbitrary scalar data.
The offset of the read or write operation specifies the starting kernel
virtual address similar to
.Pa /dev/kmem .
.It Pa ace2-ace-capability
Permits reading and writing capability addresses.
The offset of the read or write operation specifies the starting kernel
virtual address.
The length of the user request only covers the address portion of capabilities,
so on CHERI systems it will be half of the address range examined.
Reads return the addresses of capability in the requested range,
while write operations set addresses of existing capabilities in the requested
range preserving other capability metadata.
Note that setting an address of a capability may clear the capability's tag
if the new address is unrepresentable.
.It Pa ace2-ace-copycap
Each write operation copies a single capability from one kernel address to
another.
The offset of the write operation specifies the destination kernel virtual
address.
The write payload contains the source kernel virtual address
(as a binary
.Vt ptraddr_t
value) of the capability to copy.
.El
.Sh FILES
.Bl -tag -width "/dev/ace2-ace-capability" -compact
.It Pa /boot/kernel/ace2.ko
.It Pa /dev/ace2-ace-capability
.It Pa /dev/ace2-ace-copycap
.It Pa /dev/ace2-ace-data
.Sh DIAGNOSTICS
.Bl -diag
.It "ace2: ACE2_COPY_CAPABILITY from %p to %p"
A capability was copied from the first address to the second address.
.It "ace2: ACE2_READ_CAPABILITY %zu pointers at %p"
The addresses for an array of capabilities were read from the given address.
.It "ace2: ACE2_READ_CAPABILITY EFAULT %zu pointers at %p"
An attempt to read the addresses of an array of capabilities from the given
address failed.
.It "ace2: ACE2_READ_DATA %zu bytes at %p"
An array of bytes were read from the given address.
.It "ace2: ACE2_READ_DATA EFAULT %zu bytes at %p"
An attempt to read an array of bytes from the given address failed.
.It "ace2: ACE2_WRITE_CAPABILITY %zu pointers at %p"
New addresses were set for an array of capabilities at the given address.
.It "ace2: ACE2_WRITE_CAPABILITY EFAULT %zu pointers at %p"
An attempt to set the addresses for an array of capabilities at the given
address failed.
.It "ace2: ACE2_WRITE_DATA %zu bytes at %p"
An array of bytes were written to the given address.
.It "ace2: ACE2_WRITE_DATA EFAULT %zu bytes at %p"
An attempt to write an array of bytes to the given address failed.
.El
.El


sys/dev/sift/ace2/ace2.c
/*-
 * Copyright (c) 2025 John Baldwin <john@araratriver.co>
 *
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * This software was developed by SRI International, the University of
 * Cambridge Computer Laboratory (Department of Computer Science and
 * Technology), and Capabilities Limited under Defense Advanced Research
 * Projects Agency (DARPA) Contract No. FA8750-24-C-B047 ("DEC").
 */

#include <sys/param.h>
#include <sys/conf.h>
#include <sys/fcntl.h>
#include <sys/kernel.h>
#include <sys/module.h>
#include <sys/priv.h>
#include <sys/proc.h>
#include <sys/uio.h>
#include <cheri/cheric.h>
#include <vm/vm.h>
#include <vm/pmap.h>
#include <vm/vm_extern.h>
#include <vm/vm_param.h>

/*
 * Wrapper for printf() that outputs an "ace2: " prefix for each
 * message.
 */
#define	pr_sift(fmt, ...)						\
	printf("ace2: " fmt, ##__VA_ARGS__)

static d_open_t ace2_open;
static d_read_t ace2_data_rdwr;
static d_read_t ace2_capability_rdwr;
static d_write_t ace2_copycap_write;

static struct cdevsw ace2_data_cdevsw = {
	.d_version = D_VERSION,
	.d_name = "ace2_data",
	.d_open = ace2_open,
	.d_read = ace2_data_rdwr,
	.d_write = ace2_data_rdwr
};

static struct cdevsw ace2_capability_cdevsw = {
	.d_version = D_VERSION,
	.d_name = "ace2_capability",
	.d_open = ace2_open,
	.d_read = ace2_capability_rdwr,
	.d_write = ace2_capability_rdwr
};

static struct cdevsw ace2_copycap_cdevsw = {
	.d_version = D_VERSION,
	.d_name = "ace2_copycap",
	.d_open = ace2_open,
	.d_write = ace2_copycap_write
};

static int
ace2_open(struct cdev *dev, int oflags, int devtype, struct thread *td)
{
	int error = 0;

	/* Require same permissions as /dev/kmem. */
	if (oflags & FREAD)
		error = priv_check(td, PRIV_KMEM_READ);
	if (oflags & FWRITE) {
		if (error == 0)
			error = priv_check(td, PRIV_KMEM_WRITE);
		if (error == 0)
			error = securelevel_gt(td->td_ucred, 0);
	}

	return (error);
	
}

static int
validate_kva_range(vm_offset_t va, size_t len, vm_prot_t prot)
{
	vm_offset_t page_off;

	if (va + len < va)
		return (EINVAL);

	page_off = va % PAGE_SIZE;
	va = trunc_page(va);
	len += page_off;

	if (VIRT_IN_DMAP(va)) {
		if (!VIRT_IN_DMAP(va + len - 1))
			return (EFAULT);
	} else {
		if (!kernacc((void *)(uintptr_t)va, len, prot))
			return (EFAULT);
	}

	return (0);
}
	
static int
ace2_data_rdwr(struct cdev *dev, struct uio *uio, int ioflag)
{
	void *kva;
	int error;
	bool read = uio->uio_rw == UIO_READ;
	size_t len = uio->uio_resid;

	error = validate_kva_range(uio->uio_offset, len, read ? VM_PROT_READ :
	    VM_PROT_WRITE);
	if (error != 0)
		return (error);

	/*
	 * XXX: This is racey if KVA mappings change while blocked in
	 * uiomove.  For the purposes of this module, that is fine.
	 *
	 * NB: kernel_root_cap is part of the TCB and any sensible
	 * c18n policy would normally forbid direct access to it in
	 * other compartments.  The bounds on `kva` are also best-effort
	 * and may end up overly-broad.
	 */
#ifdef __CHERI_PURE_CAPABILITY__
	kva = cheri_setaddress(kernel_root_cap, uio->uio_offset);
	kva = cheri_setbounds(kva, len);
#else
	kva = (void *)(uintptr_t)uio->uio_offset;
#endif
	error = uiomove(kva, uio->uio_resid, uio);
	if (error != 0) {
		pr_sift("%s EFAULT %zu bytes at %p\n", read ? "ACE2_READ_DATA" :
		    "ACE2_WRITE_DATA", len, kva);
	} else {
		pr_sift("%s %zu bytes at %p\n", read ? "ACE2_READ_DATA" :
		    "ACE2_WRITE_DATA", len, kva);
	}
	return (error);
}

static int
ace2_capability_rdwr(struct cdev *dev, struct uio *uio, int ioflag)
{
	uintptr_t *kva;
	int error;
	bool read = uio->uio_rw == UIO_READ;
	size_t len = uio->uio_resid;
	size_t nptrs;

	if (len % sizeof(ptraddr_t) != 0 ||
	    !is_aligned(uio->uio_offset, sizeof(void *)))
		return (EINVAL);

	nptrs = len / sizeof(ptraddr_t);
	len = nptrs * sizeof(void *);
	error = validate_kva_range(uio->uio_offset, len, read ? VM_PROT_READ :
	    VM_PROT_WRITE);
	if (error != 0)
		return (error);

	/* Comments in ace2_data_rdwr apply here as well. */
#ifdef __CHERI_PURE_CAPABILITY__
	kva = cheri_setaddress(kernel_root_cap, uio->uio_offset);
	kva = cheri_setbounds(kva, len);
	for (size_t i = 0; i < nptrs; i++) {
		ptraddr_t addr;

		if (read)
			addr = cheri_getaddress(kva[i]);
		
		error = uiomove(&addr, sizeof(addr), uio);
		if (error != 0)
			break;

		if (!read)
			kva[i] = cheri_setaddress(kva[i], addr);
	}
#else
	kva = (void *)(uintptr_t)uio->uio_offset;
	error = uiomove(kva, uio->uio_resid, uio);
#endif
	if (error != 0) {
		pr_sift("%s EFAULT %zu pointers at %p\n",
		    read ? "ACE2_READ_CAPABILITY" : "ACE2_WRITE_CAPABILITY",
		    nptrs, kva);
	} else {
		pr_sift("%s %zu pointers at %p\n",
		    read ? "ACE2_READ_CAPABILITY" : "ACE2_WRITE_CAPABILITY",
		    nptrs, kva);
	}
	return (error);
}

static int
ace2_copycap_write(struct cdev *dev, struct uio *uio, int ioflag)
{
	uintptr_t *dst, *src;
	ptraddr_t addr;
	int error;

	if (uio->uio_resid != sizeof(addr) ||
	    !is_aligned(uio->uio_offset, sizeof(void *)))
		return (EINVAL);

	error = validate_kva_range(uio->uio_offset, sizeof(void *),
	    VM_PROT_WRITE);
	if (error != 0)
		return (error);

#ifdef __CHERI_PURE_CAPABILITY__
	dst = cheri_setaddress(kernel_root_cap, uio->uio_offset);
	dst = cheri_setboundsexact(dst, sizeof(void *));
#else
	dst = (void *)(uintptr_t)uio->uio_offset;
#endif

	error = uiomove(&addr, sizeof(addr), uio);
	if (error != 0)
		return (error);

	error = validate_kva_range(addr, sizeof(void *), VM_PROT_READ);
	if (error != 0)
		return (error);

#ifdef __CHERI_PURE_CAPABILITY__
	src = cheri_setaddress(kernel_root_cap, addr);
	src = cheri_setboundsexact(src, sizeof(void *));
#else
	src = (void *)(uintptr_t)addr;
#endif
	*dst = *src;
	pr_sift("ACE2_COPY_CAPABILITY from %p to %p\n", src, dst);
	return (0);
}

static int
ace2_modevent(module_t mod, int type, void *data)
{
	static struct cdev *ace2_data;
	static struct cdev *ace2_capability;
	static struct cdev *ace2_copycap;

	switch (type) {
	case MOD_LOAD:
		ace2_data = make_dev(&ace2_data_cdevsw, 0, UID_ROOT, GID_WHEEL,
		    0600, "ace2-ace-data");
		ace2_capability = make_dev(&ace2_capability_cdevsw, 0,
		    UID_ROOT, GID_WHEEL, 0600, "ace2-ace-capability");
		ace2_copycap = make_dev(&ace2_copycap_cdevsw, 0,
		    UID_ROOT, GID_WHEEL, 0600, "ace2-ace-copycap");
		return (0);
	case MOD_UNLOAD:
		destroy_dev(ace2_data);
		destroy_dev(ace2_capability);
		destroy_dev(ace2_copycap);
		return (0);
	default:
		return (EOPNOTSUPP);
	}
}

DEV_MODULE(ace2, ace2_modevent, NULL);
